File: android/app/src/main/java/com/krya1012/clearance/data/PeriodicTaskStore.kt
package com.krya1012.clearance.data

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import kotlinx.coroutines.flow.first
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.Calendar

/**
 * DataStore persistence for PeriodicTask, following the ScheduleStore pattern.
 * Completions auto-reset when `load()` is called after a week or month boundary.
 */
class PeriodicTaskStore(private val context: Context) {

    private object Keys {
        val TASKS = stringPreferencesKey("Clearance.periodicTasks.v1")
        val SEED_VERSION = intPreferencesKey("Clearance.periodicSeedVersion.v1")
    }

    companion object {
        private const val CURRENT_SEED_VERSION = 2
    }

    suspend fun seedDefaultsIfNeeded() {
        val stored = context.clearanceDataStore.data.first()[Keys.SEED_VERSION] ?: 0
        // Also reseed if the stored blob is corrupted (present but undecodable) —
        // otherwise a decode failure would silently look like "already seeded"
        // forever and the next save() would permanently overwrite it with
        // whatever the corrupted read decoded to (nothing).
        if (stored >= CURRENT_SEED_VERSION && !isStoredDataCorrupted()) return

        val existing = load()
        val existingTitles = existing.map { it.title }.toSet()
        val seeds = listOf(
            PeriodicTask(title = "Laundry", emoji = "🧺", recurrence = Recurrence.WEEKLY),
            PeriodicTask(title = "Shopping", emoji = "🛒", recurrence = Recurrence.WEEKLY),
            PeriodicTask(title = "Budget", emoji = "💰", recurrence = Recurrence.MONTHLY),
            PeriodicTask(title = "Post", emoji = "📬", recurrence = Recurrence.MONTHLY),
        )
        val toAdd = seeds.filterNot { existingTitles.contains(it.title) }
        if (toAdd.isNotEmpty()) save(existing + toAdd)
        context.clearanceDataStore.edit { it[Keys.SEED_VERSION] = CURRENT_SEED_VERSION }
    }

    suspend fun load(): List<PeriodicTask> {
        val raw = context.clearanceDataStore.data.first()[Keys.TASKS] ?: return emptyList()
        val tasks = runCatching { Json.decodeFromString<List<PeriodicTask>>(raw) }.getOrNull()
            ?: return emptyList()

        val now = Calendar.getInstance()
        val currentWeek = now.get(Calendar.WEEK_OF_YEAR)
        val currentYear = now.get(Calendar.YEAR)
        val currentMonth = now.get(Calendar.MONTH)

        return tasks.map { task ->
            if (!task.isCompleted || task.completedDateMillis == null) return@map task
            val completed = Calendar.getInstance().apply { timeInMillis = task.completedDateMillis }
            val stillCurrent = when (task.recurrence) {
                Recurrence.WEEKLY ->
                    completed.get(Calendar.WEEK_OF_YEAR) == currentWeek && completed.get(Calendar.YEAR) == currentYear
                Recurrence.MONTHLY ->
                    completed.get(Calendar.MONTH) == currentMonth && completed.get(Calendar.YEAR) == currentYear
                Recurrence.DAILY -> true
            }
            if (stillCurrent) task else task.copy(isCompleted = false, completedDateMillis = null)
        }
    }

    suspend fun save(tasks: List<PeriodicTask>) {
        val encoded = Json.encodeToString(tasks)
        context.clearanceDataStore.edit { it[Keys.TASKS] = encoded }
    }

    /**
     * True when a periodic-tasks blob exists but fails to decode — distinct
     * from "no data yet" (key absent) or "user cleared everything" (decodes
     * fine to an empty list), neither of which should trigger a reseed.
     */
    private suspend fun isStoredDataCorrupted(): Boolean {
        val raw = context.clearanceDataStore.data.first()[Keys.TASKS] ?: return false
        return runCatching { Json.decodeFromString<List<PeriodicTask>>(raw) }.getOrNull() == null
    }
}

File: ios/Clearance/Helpers/PeriodicTaskStore.swift
//
//  PeriodicTaskStore.swift
//  Clearance
//
//  UserDefaults persistence for PeriodicTask, following the ScheduleStore pattern.
//  Completions auto-reset when load() is called after a week or month boundary.
//

import Foundation

@MainActor
final class PeriodicTaskStore {

    private let defaults: UserDefaults
    private static let key = "Clearance.periodicTasks.v1"
    private static let seedVersionKey = "Clearance.periodicSeedVersion.v1"
    private static let currentSeedVersion = 2

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    func seedDefaultsIfNeeded() {
        let stored = defaults.integer(forKey: Self.seedVersionKey)
        guard stored < Self.currentSeedVersion else { return }
        let existing = load()
        let existingTitles = Set(existing.map(\.title))
        let seeds: [PeriodicTask] = [
            PeriodicTask(title: "Laundry",  emoji: "🧺", recurrence: .weekly),
            PeriodicTask(title: "Shopping", emoji: "🛒", recurrence: .weekly),
            PeriodicTask(title: "Budget",   emoji: "💰", recurrence: .monthly),
            PeriodicTask(title: "Post",     emoji: "📬", recurrence: .monthly),
        ]
        let toAdd = seeds.filter { !existingTitles.contains($0.title) }
        if !toAdd.isEmpty { save(existing + toAdd) }
        defaults.set(Self.currentSeedVersion, forKey: Self.seedVersionKey)
    }

    func load() -> [PeriodicTask] {
        guard
            let data = defaults.data(forKey: Self.key),
            var tasks = try? JSONDecoder().decode([PeriodicTask].self, from: data)
        else { return [] }

        let cal = Calendar.current
        let now = Date()
        let currentWeek  = cal.component(.weekOfYear, from: now)
        let currentYear  = cal.component(.year, from: now)
        let currentMonth = cal.component(.month, from: now)

        for i in tasks.indices {
            guard tasks[i].isCompleted, let completedDate = tasks[i].completedDate else { continue }
            switch tasks[i].recurrence {
            case .weekly:
                let w = cal.component(.weekOfYear, from: completedDate)
                let y = cal.component(.year, from: completedDate)
                if w != currentWeek || y != currentYear {
                    tasks[i].isCompleted = false
                    tasks[i].completedDate = nil
                }
            case .monthly:
                let m = cal.component(.month, from: completedDate)
                let y = cal.component(.year, from: completedDate)
                if m != currentMonth || y != currentYear {
                    tasks[i].isCompleted = false
                    tasks[i].completedDate = nil
                }
            case .daily:
                break
            }
        }

        return tasks
    }

    func save(_ tasks: [PeriodicTask]) {
        guard let data = try? JSONEncoder().encode(tasks) else { return }
        defaults.set(data, forKey: Self.key)
    }
}
